stlab/channel.hpp

```diff
@@ -352,17 +388,24 @@ struct shared_process : shared_process_receiver<yield_type<T, Arg>>,
     }
 
     void clear_to_send() override {
+        // The downstream sender may trigger this because an other parallel upstream
+        // process may have just passed down a value
+        if (_process_suspend_count == 0)
```

Comment:
I understand your point, but there two possibilities from my point of view to solve the join problem:
- One collects all values in a common/shared context and then the last upstream, that received the value, passes this down to the joining process. Then the sending upstream process is in suspend state and would be unsuspended by the joining process. But up to now the joining process does not know which upstream send the value and so all has to be informed with a cts. Since only the sending upstream was suspended, all the other upstream would complain that they were not suspended before.
- The other solutions would be, that each upstream part sends its value dirctly to the joining process and that yields when all values have arrived. But the dequeue() of the joining process would trigger a cts() on all upstreamer whenever the _process_message_queue runs dry. And so again we run into the problem, that all the upstreamer that just did not send a value, are not in suspend state.

At the moment I understood the design of the channels in such a way, that no downstream has any inside knowledge about their connected upstreamer. Which I think is good.

An other option could be, that one passes a pair<> to the downstream, that holds the value as first element and a function object that contains a call to cts as second element. So only that upstream would get called with cts, that actually send the value. 

What are your ideas on this?